dali/operators/math/expressions/broadcasting.h

```diff
@@ -0,0 +1,128 @@
+ *        with extend 1.
+ *        Example: lhs=(10, 10, 3), rhs=(1, 3) -> result=(10, 10, 3)
+ * 
+ * @param result resulting shape (should as many dimensions as the operands)
+ * @param lhs shape of left hand side operand
+ * @param rhs shape of right hand side operand
+ * @return TensorShape<> resulting shape
+ */
+DLL_PUBLIC void BroadcastShape(TensorShape<>& result, span<const TensorShape<>> shapes);
+DLL_PUBLIC void BroadcastShape(TensorListShape<>& result, span<const TensorListShape<>> shapes);
+
+template <typename Shape>
+void BroadcastShape(Shape &result, const Shape& a, const Shape& b) {
+  std::array<Shape, 2> arr = {a, b};
+  BroadcastShape(result, make_cspan(arr));
+}
+
+template <typename Shape>
+void BroadcastShape(Shape &result, const Shape &a, const Shape &b, const Shape &c) {
+  std::array<Shape, 3> arr = {a, b, c};
+  BroadcastShape(result, make_cspan(arr));
+}
+
+/**
+ * @brief Verifies whether two shapes or more can be broadcasted
+ *        Two shapes can be broadcasted if all the extents are either 
+ *        equal or one of them is equal to one for the `ndim` rightmost
+ *        dimensions, being `ndim` the minimum number of dimensions of 
+ *        the two.
+ *        If the shapes are identical (no broadcasting needed), true will
+ *        be returned.
+ */
+DLL_PUBLIC bool CanBroadcast(span<const TensorShape<>> shapes);
+DLL_PUBLIC bool CanBroadcast(span<const TensorListShape<>> shapes);
+
+template <typename Shape>
+bool CanBroadcast(const Shape &a, const Shape &b) {
+  std::array<Shape, 2> arr = {a, b};
+  return CanBroadcast(make_cspan(arr));
+}
+
+template <typename Shape>
+bool CanBroadcast(const Shape &a, const Shape &b, const Shape& c) {
+  std::array<Shape, 3> arr = {a, b, c};
+  return CanBroadcast(make_cspan(arr));
+}
+
+
+/**
+ * @brief Returns true if the shapes require broadcasting
+ *        Two or more shapes require broadcasting if they are not scalar-like
+ *        and their shapes are not equal.
+ * @remarks This function does not check whether the shapes can be broadcasted
+ *          For this, use CanBroadcast
+ */
+DLL_PUBLIC bool NeedBroadcasting(span<const TensorShape<>> shapes);
+DLL_PUBLIC bool NeedBroadcasting(span<const TensorListShape<>> shapes);
+
+template <typename Shape>
+bool NeedBroadcasting(const Shape &a, const Shape &b) {
+  std::array<Shape, 2> arr = {a, b};
+  return NeedBroadcasting(make_cspan(arr));
+}
+
+template <typename Shape>
+bool NeedBroadcasting(const Shape &a, const Shape &b, const Shape& c) {
+  std::array<Shape, 3> arr = {a, b, c};
+  return NeedBroadcasting(make_cspan(arr));
+}
+
+/**
+ * @brief Calculates strides to cover a possibly broadcasted shape. 
+ *        The stride for those broadcasted dimensions is set to 0
+ */
+DLL_PUBLIC TensorShape<> StridesForBroadcasting(const TensorShape<> &out_sh,
+                                                const TensorShape<> &in_sh,
+                                                const TensorShape<> &in_strides);
```

Comment:
Just curious, why do we have in_sh and in_strides here instead of computing the strides on the fly?

Reply:
Imagine that we were doing ROI calculation. The strides will remain as the ones of the original image, but the shape will be the one of the ROI. That was my thinking to provide both.